framework/doc/content/source/kernels/ConservativeAdvection.md

```diff
@@ -22,27 +22,104 @@ element solution of the weak formulation. The first term is the volumetric term
 is a surface term describing the advective flux out of the
 volume. `ConservativeAdvection` corresponds to the former volumetric term.
 
-The corresponding Jacobian is given by
+Without numerical stabilization the corresponding Jacobian is given by
 
 \begin{equation}
 \frac{\partial R_i(u_h)}{\partial u_j} = -(\nabla \psi_i, \vec{v} \phi_j).
 \end{equation}
 
+## Full upwinding
+
+Advective flow is notoriously prone to physically-incorrect overshoots
+and undershoots, so in many simulations some numerical stabilization
+is used to reduce or eliminate this spurious behaviour.
+Full-upwinding [cite:dalen1979,adhikary2011] is an example of
+numerical stabilization and this essentially adds numerical diffusion
+to completely eliminate overshoots and undershoots.  Full-upwinding is
+available in `ConservativeAdvection` by setting the `upwinding_type`
+appropriately.
+
+!alert note
+Full upwinding only works for continuous FEM
+
+For each element in the mesh, full upwinding is implemented in the
+following way.  For each $i$, the quantity
+\begin{equation}
+\tilde{R}_{i} = -(\nabla\psi_i, \vec{v}) \ ,
+\end{equation}
+is evaluated.  If $\tilde{R}_{i}>0$ is positive then mass (or heat, or whatever $u$
+represents) is flowing *out* of node $i$, and this is called an
+"upwind" node.  If $\tilde{R}_{i}\geq 0$, the residual for node $i$
+is set to
+\begin{equation}
+R_{i} = u_{i}\tilde{R}_{i} \ \ \texttt{ for } \ \ \tilde{R}_{i}\geq 0.
+\end{equation}
+Here $u_{i}$ is the value of $u$ at the node $i$.  Define the total
+mass flowing from the upwind nodes:
+\begin{equation}
+M_{\mathrm{out}} = \sum_{i\ \mathrm{ with }\ \tilde{R}_{i}\geq 0}
+u_{i}\tilde{R}_{i} \ .
+\end{equation}
+Similarly, define
+\begin{equation}
+\tilde{M}_{\mathrm{in}} = - \sum_{i \ \mathrm{ with }\  \tilde{R}_{i}< 0}
+\tilde{R}_{i} \ .
+\end{equation}
+Mass is conserved if the residual for the downwind nodes is defined to
+be
+\begin{equation}
+R_{i} = \tilde{R}_{i} M_{\mathrm{out}} / M_{\mathrm{in}}  \ \ \texttt{ for } \ \ \tilde{R}_{i}< 0.
+\end{equation}
```

Comment:
I should have asked this: this treatment works with the case where other terms exist like the diffusion or source term, right? I knew SUPG cannot be done just on one term. Maybe a small paragraph on discussing the advantage of this scheme will answer readers' questions like this.

Reply:
Yes, you may use full upwinding on just the advective term if you have other terms (like diffusion).  It is a problem-specific choice, really.  If your diffusion (etc) terms are complicated nonlinear things you may choose to fully-upwind them too but you don't *need* to.

Reply:
Addressed: added a paragraph to doco.